### PR TITLE
Fix various problems in GDML writing with Geant4 units

### PR DESCRIPTION
**Pull request by MarkusFrankATcernch**
# This Pull request:
- Allow users to specify the floating point precision of the GDML attributes and elements
- Fix position units: take the chosen default length units rather than default argument 
  from member function which was "cm"
- Eject material property refs as first child elements as required from the GDML schema
- Add the atomic number N to the `<atom>` attributes.
- Ensure proper NCN names for surfaces (No '/' and '#' characters) for surfaces
- Eject **const properties** as matrices in GDML with one element rather than constants.
- Make physical volume names unique as required by the GDML schema.
  This is also mandatory to properly define surfaces.
- Exclude the "dummy" material from being written to the GDML file.
  The dummy material is internally used by TGeo, but not understood by Geant4.
  Requires the new option `TGDMLWrite::SetIgnoreDummyMaterial(true)` to be set by the user.
- Reading the produced GDML files back with Geant4 was tested.

Backwards compatibility should be preserved 100 %

## Checklist:

- [ x] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by agheata** [comment]
@MarkusFrankATcernch I'm checking this today, I'm still seeing mixes of different features in some commits. Ideally, the different commits will treat completely different features, and a test should be present where possible.

File: geom/gdml/src/TGDMLParse.cxx [hunk 1]
```diff
@@ -594,15 +594,20 @@ XMLNodePointer_t TGDMLParse::MatrixProcess(TXMLEngine *gdml, XMLNodePointer_t no
...
+   }
+   else  {
+     TGDMLMatrix *matrix = new TGDMLMatrix(name, valueList.size() / coldim, coldim);
+     matrix->SetMatrixAsString(values.c_str());
+     for (size_t i = 0; i < valueList.size(); ++i)
+       matrix->Set(i / coldim, i % coldim, valueList[i]);
+     
+     gGeoManager->AddGDMLMatrix(matrix);
+     fmatrices[name.Data()] = matrix;
+   }
```
Comment (agheata) [inline geom/gdml/src/TGDMLParse.cxx:610]: This part seems unrelated to ignoring dummy materials

File: geom/gdml/src/TGDMLWrite.cxx [hunk 2]
```diff
@@ -1873,9 +1886,13 @@ XMLNodePointer_t TGDMLWrite::CreateBorderSurfaceN(TGeoBorderSurface * geoSurf)
...
    fGdmlE->NewAttr(mainN, nullptr, "name", geoSurf->GetName());
    fGdmlE->NewAttr(mainN, nullptr, "surfaceproperty", geoSurf->GetTitle());
-   // Cretate the logical volume reference node
+
+   // Cretate the logical volume reference nodes
    auto childN = fGdmlE->NewChild(nullptr, nullptr, "physvolref", nullptr);
    fGdmlE->NewAttr(childN, nullptr, "ref", geoSurf->GetNode1()->GetName());
+   fGdmlE->AddChild(mainN, childN);
+
+   childN = fGdmlE->NewChild(nullptr, nullptr, "physvolref", nullptr);
```
Comment (agheata) [inline geom/gdml/src/TGDMLWrite.cxx:1914]: Also unrelated to the treatment of dummy materials, can you please separate functionally the commits, otherwise it is going to be very difficult to track/debug the changes

File: geom/gdml/src/TGDMLWrite.cxx [hunk 3]
```diff
@@ -1938,10 +1955,11 @@ XMLNodePointer_t TGDMLWrite::CreateMatrixN(TGDMLMatrix const *matrix)
 
 XMLNodePointer_t TGDMLWrite::CreateConstantN(const char *name, Double_t value)
 {
-   XMLNodePointer_t mainN = fGdmlE->NewChild(nullptr, nullptr, "constant", nullptr);
+   XMLNodePointer_t mainN = fGdmlE->NewChild(nullptr, nullptr, "matrix", nullptr);
    const TString fltPrecision = TString::Format("%%.%dg", fFltPrecision);
    fGdmlE->NewAttr(mainN, nullptr, "name", name);
-   fGdmlE->NewAttr(mainN, nullptr, "value", TString::Format(fltPrecision.Data(), value));
+   fGdmlE->NewAttr(mainN, nullptr, "coldim", "1");
+   fGdmlE->NewAttr(mainN, nullptr, "values", TString::Format(fltPrecision.Data(), value));
```
Comment (agheata) [inline geom/gdml/src/TGDMLWrite.cxx:1984]: Same as above, unrelated

File: geom/gdml/src/TGDMLWrite.cxx [hunk 4]
```diff
@@ -527,12 +527,14 @@ XMLNodePointer_t TGDMLWrite::ExtractMaterials(TList* materialsLst)
...
+   TGeoMedium   *dummy_med = TGeoVolume::DummyMedium();
+   TGeoMaterial *dummy_mat = dummy_med ? dummy_med->GetMaterial() : nullptr;
 
    while ((lmaterial = (TGeoMaterial *)next())) {
       //generate uniq name
       TString mname = lmaterial->GetName();
-      if ( fIgnoreDummyMaterial && mname == "dummy" )   {
-        Info("ExtractMaterials", "Skip dummy material!");
+      if ( fIgnoreDummyMaterial && dummy_mat && dummy_mat->GetName() == mname )   {
+        Info("ExtractMaterials", "Skip dummy material: %s", dummy_mat->GetName());
```
Comment (agheata) [inline geom/gdml/src/TGDMLWrite.cxx]: These changes should appear in the commit about dummy materials

File: geom/gdml/src/TGDMLWrite.cxx [hunk 5]
```diff
@@ -529,15 +546,15 @@ XMLNodePointer_t TGDMLWrite::ExtractMaterials(TList* materialsLst)
...
    while ((lmaterial = (TGeoMaterial *)next())) {
       //generate uniq name
-      TString mname = lmaterial->GetName();
-      if ( fIgnoreDummyMaterial && dummy_mat && dummy_mat->GetName() == mname )   {
+      std::string mname = lmaterial->GetName();
+      if ( fIgnoreDummyMaterial && dummy_mat && dummy_nam == mname )   {
         Info("ExtractMaterials", "Skip dummy material: %s", dummy_mat->GetName());
         continue;
       }
-      TString lname = GenName(mname, TString::Format("%p", lmaterial));
```
Comment (agheata) [inline geom/gdml/src/TGDMLWrite.cxx]: Should be in the dummy materials commit

File: geom/gdml/src/TGDMLWrite.cxx [hunk 6]
```diff
@@ -869,6 +890,24 @@ XMLNodePointer_t TGDMLWrite::CreateMixtureN(TGeoMixture * mixture, XMLNodePointe
...
+   }
+   // Write CONST properties
+   TList const &const_properties = mixture->GetConstProperties();
+   if (const_properties.GetSize()) {
+      TIter next(&const_properties);
+      TNamed *property;
+      while ((property = (TNamed*)next()))
+        fGdmlE->AddChild(mainN, CreatePropertyN(*property));
+   }
+
```
Comment (agheata) [inline geom/gdml/src/TGDMLWrite.cxx:898]: The mixture properties fix should be in a separate commit

File: geom/gdml/src/TGDMLWrite.cxx [hunk 7]
```diff
@@ -169,6 +169,23 @@ ClassImp(TGDMLWrite);
...
+  }
+  
+  // Create a NCN compliant name (no '/' and no '#')  
+  std::string make_NCName(const std::string& in)   {
+    std::string res = str_replace(in, "/", "_");
+    res = str_replace(res, "#", "_");
+    return res;
+  }
+
+  // Materials extractor from a volume tree
```
Comment (agheata) [inline geom/gdml/src/TGDMLWrite.cxx]: All changes related to naming correction should be in a separate commit

**Review by agheata — CHANGES_REQUESTED** [review]
@MarkusFrankATcernch can you please reorganize your changes in commits holding changes for one feature at a time?

**Comment by MarkusFrankATcernch** [comment]
Better then start from scratch....

**Comment by agheata** [comment]
> Better then start from scratch....

Thanks @MarkusFrankATcernch !
